### PR TITLE
Aligned the Builder classes better with Fabric8

### PR DESCRIPTION
**Pull request by ampie**
Also upgraded to Fabric8 4.4.1
Test coverage now in 80's
Also introduced integrated tests because some serialization issues still occur

### CODE REVIEW

**Comment by entando-service** [comment]
All is well

File: src/main/java/org/entando/kubernetes/model/plugin/DoneableEntandoPlugin.java [hunk 1]
```diff
@@ -1,41 +1,49 @@
...
-    public DoneableEntandoPlugin(EntandoPlugin resource, Function function) {
-        super(resource, function);
-        this.resource = resource;
+    public DoneableEntandoPlugin(EntandoPlugin resource, Function<EntandoPlugin, EntandoPlugin> function) {
+        super(resource.getSpec(), resource.getMetadata());
+        this.function = function;
+        this.entandoStatus = Optional.ofNullable(resource.getStatus()).orElse(new EntandoCustomResourceStatus());
     }
 
     public DoneableEntandoPlugin addNewConnectionConfigName(String name) {
```
Comment (werbth) [inline src/main/java/org/entando/kubernetes/model/plugin/DoneableEntandoPlugin.java]: This public method doesn't seem to be covered by Unit Test.

File: src/main/java/org/entando/kubernetes/model/plugin/EntandoPlugin.java [hunk 2]
```diff
@@ -13,6 +14,7 @@
 @JsonDeserialize()
 public class EntandoPlugin extends CustomResource implements EntandoCustomResource, RequiresKeycloak {
 
+    public static final String CRD_NAME = "entandoplugins.entando.org";
```
Comment (werbth) [inline src/main/java/org/entando/kubernetes/model/plugin/EntandoPlugin.java]: Add blank line after this field declaration.

File: src/test/java/org/entando/kubernetes/model/AbstractEntandoAppTest.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+        }
+        return (CustomResourceOperationsImpl<EntandoApp, EntandoAppList, DoneableEntandoApp>) client
+                .customResources(entandoAppCrd, EntandoApp.class, EntandoAppList.class, DoneableEntandoApp.class);
+    }
+
+    @BeforeEach
+    public void deleteExternalDatabase() throws InterruptedException {
+        entandoApps().inNamespace(MY_NAMESPACE).withName(MY_APP).delete();
+        while (entandoApps().inNamespace(MY_NAMESPACE).list().getItems().size() > 0) {
```
Comment (werbth) [inline src/test/java/org/entando/kubernetes/model/AbstractEntandoAppTest.java]: It's better to use isEmpty to check if the collection is empty or not.

File: src/test/java/org/entando/kubernetes/model/AbstractEntandoPluginTest.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            }
+        }
+        return (CustomResourceOperationsImpl<EntandoPlugin, EntandoPluginList, DoneableEntandoPlugin>) client
+                .customResources(entandoPluginCrd, EntandoPlugin.class, EntandoPluginList.class, DoneableEntandoPlugin.class);
+    }
+
+    @BeforeEach
+    public void deleteEntandPlugin() throws InterruptedException {
+        entandoPlugins().inNamespace(MY_NAMESPACE).withName(MY_PLUGIN).delete();
+        while (entandoPlugins().inNamespace(MY_NAMESPACE).list().getItems().size() > 0) {
```
Comment (werbth) [inline src/test/java/org/entando/kubernetes/model/AbstractEntandoPluginTest.java]: It's better to use isEmpty to check if the collection is empty or not.

File: src/test/java/org/entando/kubernetes/model/AbstractExternalDatabaseTest.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+        }
+        return (CustomResourceOperationsImpl<ExternalDatabase, ExternalDatabaseList, DoneableExternalDatabase>) client
+                .customResources(externalDatabaseCrd, ExternalDatabase.class, ExternalDatabaseList.class, DoneableExternalDatabase.class);
+    }
+
+    @BeforeEach
+    public void deleteExternalDatabase() throws InterruptedException {
+        externalDatabases().inNamespace(MY_NAMESPACE).withName(MY_EXTERNAL_DATABASE).delete();
+        while (externalDatabases().inNamespace(MY_NAMESPACE).list().getItems().size() > 0) {
```
Comment (werbth) [inline src/test/java/org/entando/kubernetes/model/AbstractExternalDatabaseTest.java]: It's better to use isEmpty to check if the collection is empty or not.

File: src/test/java/org/entando/kubernetes/model/AbstractKeycloakServerTest.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+        }
+        return (CustomResourceOperationsImpl<KeycloakServer, KeycloakServerList, DoneableKeycloakServer>) client
+                .customResources(keycloakServerCrd, KeycloakServer.class, KeycloakServerList.class, DoneableKeycloakServer.class);
+    }
+
+    @BeforeEach
+    public void deleteKeycloakServer() throws InterruptedException {
+        keycloakServers().inNamespace(MY_NAMESPACE).withName(MY_KEYCLOAK).delete();
+        while (keycloakServers().inNamespace(MY_NAMESPACE).list().getItems().size() > 0) {
```
Comment (werbth) [inline src/test/java/org/entando/kubernetes/model/AbstractKeycloakServerTest.java]: It's better to use isEmpty to check if the collection is empty or not.

**Comment by entando-service** [comment]
All is well

File: pom.xml [hunk 7]
```diff
@@ -67,6 +67,7 @@
         </maven.release.repository.url>
         <org.slf4j.simpleLogger.log.io.fabric8>debug</org.slf4j.simpleLogger.log.io.fabric8>
         <org.slf4j.simpleLogger.log.org.jboss.weld>debug</org.slf4j.simpleLogger.log.org.jboss.weld>
+        <fabric8.version>4.4.1</fabric8.version>
```
[the remark this answers is not in the document]
Comment (ampie) [reply pom.xml]: Yes unfortunately the Microbeans project we use has not been keeping up. That is why we need to start using JVM Operators. Sadly I actually had to downgrade again to 4.1.2. This does however bump up the priority of our migration to JVM Operators.

File: src/test/java/org/entando/kubernetes/model/AbstractEntandoAppTest.java (hunk 3 above)
Comment (ampie) [reply src/test/java/org/entando/kubernetes/model/AbstractEntandoAppTest.java]: Nope that would result in a negative check in the while statement

File: src/test/java/org/entando/kubernetes/model/AbstractEntandoPluginTest.java (hunk 4 above)
Comment (ampie) [reply src/test/java/org/entando/kubernetes/model/AbstractEntandoPluginTest.java]: Nope that would result in a negative check in the while statement

File: src/test/java/org/entando/kubernetes/model/AbstractExternalDatabaseTest.java (hunk 5 above)
Comment (ampie) [reply src/test/java/org/entando/kubernetes/model/AbstractExternalDatabaseTest.java]: Nope that would result in a negative check in the while statement

File: src/test/java/org/entando/kubernetes/model/AbstractKeycloakServerTest.java (hunk 6 above)
Comment (ampie) [reply src/test/java/org/entando/kubernetes/model/AbstractKeycloakServerTest.java]: Nope that would result in a negative check in the while statement

File: src/main/java/org/entando/kubernetes/model/plugin/DoneableEntandoPlugin.java (hunk 1 above)
Comment (ampie) [reply src/main/java/org/entando/kubernetes/model/plugin/DoneableEntandoPlugin.java]: Yeah I deleted this method. It sits on the wrong class. I was just worried about your code needing it but if you're fine with it just rather use 'editSpec().addNewConnectionConfigName'

File: src/main/java/org/entando/kubernetes/model/keycloakserver/KeycloakServer.java [hunk 8]
```diff
@@ -13,6 +14,7 @@
 @JsonDeserialize
 public class KeycloakServer extends CustomResource implements HasIngress {
 
+    public static final String CRD_NAME = "entandokeycloakservers.entando.org";
```
[the remark this answers is not in the document]
Comment (ampie) [reply src/main/java/org/entando/kubernetes/model/keycloakserver/KeycloakServer.java]: Done

File: src/main/java/org/entando/kubernetes/model/app/EntandoApp.java [hunk 9]
```diff
@@ -14,6 +15,7 @@
 @JsonDeserialize
 public class EntandoApp extends CustomResource implements HasIngress, RequiresKeycloak {
 
+    public static final String CRD_NAME = "entandoapps.entando.org";
```
[the remark this answers is not in the document]
Comment (ampie) [reply src/main/java/org/entando/kubernetes/model/app/EntandoApp.java]: Done

File: src/main/java/org/entando/kubernetes/model/externaldatabase/ExternalDatabase.java [hunk 10]
```diff
@@ -12,6 +13,7 @@
 @JsonDeserialize
 public class ExternalDatabase extends CustomResource implements EntandoCustomResource {
 
+    public static final String CRD_NAME = "externaldatabases.entando.org";
```
[the remark this answers is not in the document]
Comment (ampie) [reply src/main/java/org/entando/kubernetes/model/externaldatabase/ExternalDatabase.java]: Done

File: src/main/java/org/entando/kubernetes/model/plugin/EntandoPlugin.java (hunk 2 above)
Comment (ampie) [reply src/main/java/org/entando/kubernetes/model/plugin/EntandoPlugin.java]: Done

**Comment by entando-service** [comment]
All is well

File: src/main/java/org/entando/kubernetes/model/EntandoCustomResourceStatus.java [hunk 11]
```diff
@@ -31,41 +34,19 @@ public void setEntandoDeploymentPhase(EntandoDeploymentPhase entandoDeploymentPh
...
-        return dbServerStatus;
-    }
-
-    public void setDbServerStatus(List<DbServerStatus> dbServerStatus) {
-        this.dbServerStatus = dbServerStatus;
-    }
-
-    public List<WebServerStatus> getWebServerStatuses() {
-        return webServerStatuses;
+        return serverStatuses.values().stream().anyMatch(s -> s.hasFailed());
```
[the remark this answers is not in the document]
Comment (ampie) [reply src/main/java/org/entando/kubernetes/model/EntandoCustomResourceStatus.java]: Done

**Comment by entando-service** [comment]
All is well

**Comment by entando-service** [comment]
Something went wrong. Investigate!

**Comment by entando-service** [comment]
All is well

**Comment by entando-service** [comment]
All is well

**Pull request merged**